File: packages/icoder-python/icoder_sdk/resources/oauth.py
"""OAuth resource."""

from typing import Optional
from urllib.parse import quote

from ..client import iCoDerClient
from ..request_options import RequestOptions


class OAuthResource:
    def __init__(self, client: iCoDerClient):
        self._client = client

    def get_token(
        self,
        client_id: str,
        client_secret: str,
        request_options: Optional[RequestOptions] = None,
    ) -> dict:
        resp = self._client.post("/api/oauth/token", data={
            "client_id": client_id, "client_secret": client_secret,
            "grant_type": "client_credentials",
            "scope": "api:read api:write",
        }, request_options=request_options)
        resp.raise_for_status()
        return resp.json()

    def create_client(
        self,
        name: str,
        description: str,
        scopes: str,
        *,
        allowed_agent_ids: list[str] | None = None,
        allowed_purposes: list[str] | None = None,
        request_options: Optional[RequestOptions] = None,
    ) -> dict:
        resp = self._client.post("/api/oauth/clients", data={
            "name": name,
            "description": description,
            "scopes": scopes,
            "allowed_agent_ids": ",".join(allowed_agent_ids or []),
            "allowed_purposes": ",".join(allowed_purposes or []),
        }, request_options=request_options)
        resp.raise_for_status()
        return resp.json()

    def list_clients(
        self, request_options: Optional[RequestOptions] = None,
    ) -> dict:
        resp = self._client.get(
            "/api/oauth/clients", request_options=request_options,
        )
        resp.raise_for_status()
        return resp.json()

    def update_delegation(
        self,
        client_id: str,
        *,
        allowed_agent_ids: list[str],
        allowed_purposes: list[str],
        request_options: Optional[RequestOptions] = None,
    ) -> dict:
        resp = self._client.patch(
            f"/api/clients/{quote(client_id, safe='')}/delegation",
            json={
                "allowed_agent_ids": allowed_agent_ids,
                "allowed_purposes": allowed_purposes,
            },
            request_options=request_options,
        )
        resp.raise_for_status()
        return resp.json()

    def revoke_client(
        self,
        client_id: str,
        request_options: Optional[RequestOptions] = None,
    ) -> None:
        resp = self._client.delete(
            f"/api/oauth/clients/{quote(client_id, safe='')}",
            request_options=request_options,
        )
        resp.raise_for_status()

File: packages/icoder-sdk/src/resources/oauth.ts
// OAuth resource
import type { AxiosInstance } from 'axios';
import type { TokenResponse } from '../types.js';
import { requestConfig, type iCoDerRequestOptions } from '../request-options.js';

export class OAuthResource {
  constructor(private http: AxiosInstance) {}

  async getToken(
    clientId: string,
    clientSecret: string,
    options?: iCoDerRequestOptions,
  ): Promise<TokenResponse> {
    const form = new URLSearchParams({
      grant_type: 'client_credentials',
      client_id: clientId,
      client_secret: clientSecret,
      scope: 'api:read api:write',
    });
    const { data } = await this.http.post<TokenResponse>(
      '/api/oauth/token',
      form,
      requestConfig(
        options,
        {},
        { 'Content-Type': 'application/x-www-form-urlencoded' },
      ),
    );
    return data;
  }

  async createClient(
    name: string,
    description: string,
    scopes: string,
    delegation: { allowedAgentIds?: string[]; allowedPurposes?: string[] } = {},
    options?: iCoDerRequestOptions,
  ): Promise<{ client_id: string; client_secret: string; allowed_agent_ids?: string[]; allowed_purposes?: string[] }> {
    const form = new URLSearchParams({
      name,
      description,
      scopes,
      allowed_agent_ids: (delegation.allowedAgentIds ?? []).join(','),
      allowed_purposes: (delegation.allowedPurposes ?? []).join(','),
    });
    const { data } = await this.http.post(
      '/api/oauth/clients',
      form,
      requestConfig(
        options,
        {},
        { 'Content-Type': 'application/x-www-form-urlencoded' },
      ),
    );
    return data;
  }

  async listClients(options?: iCoDerRequestOptions): Promise<{ clients: any[] }> {
    const { data } = await this.http.get('/api/oauth/clients', requestConfig(options));
    return data;
  }

  async updateDelegation(
    clientId: string,
    allowedAgentIds: string[],
    allowedPurposes: string[],
    options?: iCoDerRequestOptions,
  ): Promise<Record<string, unknown>> {
    const { data } = await this.http.patch(
      `/api/clients/${encodeURIComponent(clientId)}/delegation`,
      { allowed_agent_ids: allowedAgentIds, allowed_purposes: allowedPurposes },
      requestConfig(options),
    );
    return data;
  }

  async revokeClient(clientId: string, options?: iCoDerRequestOptions): Promise<void> {
    await this.http.delete(
      `/api/oauth/clients/${encodeURIComponent(clientId)}`,
      requestConfig(options),
    );
  }
}
